### PR TITLE
Add concurrent upload and downloads source parameters

### PR DESCRIPTION
**Pull request by drahnr**
Minimal changes to address the root issue of https://github.com/concourse/docker-image-resource/issues/143

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@drahnr Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/docker-image-resource&pullRequestId=144)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=concourse/docker-image-resource&pullRequestId=144) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@drahnr Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/docker-image-resource&pullRequestId=144)!

File: assets/in [hunk 1]
```diff
@@ -32,6 +32,8 @@ repository="$(jq -r '.source.repository // ""' < $payload)"
 tag="$(jq -r '.source.tag // "latest"' < $payload)"
 ca_certs=$(jq -r '.source.ca_certs // []' < $payload)
 client_certs=$(jq -r '.source.client_certs // []' < $payload)
+max_concurrent_downloads=$(jq -r '.source.max_concurrent_downloads // 3' < $payload || echo "3")
+max_concurrent_uploads=$(jq -r '.source.max_concurrent_uploads // 3' < $payload || echo "3")
```
Comment (jtarchie) [inline assets/in]: Why the `echo 3`?

Comment (drahnr) [reply assets/in]: Default value 3

Comment (jtarchie) [reply assets/in]: Indeed it [does](https://stedolan.github.io/jq/manual/#Alternativeoperator://).

**Comment by drahnr** [comment]
I removed it, anything else necessary?

File: assets/out [hunk 2]
```diff
@@ -33,6 +33,8 @@ password=$(jq -r '.source.password // ""' < $payload)
 repository=$(jq -r '.source.repository // ""' < $payload)
 ca_certs=$(jq -r '.source.ca_certs // []' < $payload)
 client_certs=$(jq -r '.source.client_certs // []' < $payload)
+max_concurrent_downloads=$(jq -r '.source.max_concurrent_downloads // 3' < $payload || echo "3")
```
Comment (jtarchie) [inline assets/out]: Can you do the same as above?

Comment (jtarchie) [reply assets/out]: Is three the default the docker always provides? This is backwards change that would make other's pipelines go slower until they updated to max a higher max. Maybe we should make this optional, as in, you provide a value use it, no value, don't set any min/max.

Comment (drahnr) [reply assets/out]: Yes, docker default for download is 3, upload is - for no obvious reason - defaulting to 5

Comment (vito) [reply assets/out]: Is the `|| echo "3"` necessary? The `// 3` should default the value to `3` if it's not present.

Comment (drahnr) [reply assets/out]: No it is not but it has been removed in https://github.com/concourse/docker-image-resource/pull/144/commits/99994f0fc18f7d7e72f713013df4034e41e30c1b

**Comment by drahnr** [comment]
This is a trivial change, so can we please move forward or let me know what is necessary to change in order to get it merged or drop it altogether :)

**Comment by vito** [comment]
@drahnr Hi, sorry for the wait on this - there's a lot of tech debt in this resource so we've been hesitant to process PRs without tests, and most of them don't have any since part of the debt is that there *aren't any* to reference when submitting a PR. :)

We've just finished backfilling some around the `dockerd` command with [this commit](https://github.com/concourse/docker-image-resource/commit/5edc3b8f587ff028c7db887841865b17037b4143) - could you try using it as a reference for adding tests for this PR?

p.s. the `|| echo "3"` was indeed removed from `/in` but not `/out`. 😄

**Comment by drahnr** [comment]
Oops I totally missed that. Thanks! 

Will add a few tests next weekend.

**Comment by drahnr** [comment]
I totally forgot about this, should be done with 42ab26f.

**Comment by drahnr** [comment]
I totally forgot about this, 35e9968 should complete it.

**Comment by vito** [comment]
@drahnr No worries! Another *poke* for tests and a rebase? :) (See https://github.com/concourse/docker-image-resource/blob/master/tests/out_test.go for where to add them now.)

**Comment by drahnr** [comment]
@vito hopefully this suffices now, I also squashed my commits into one

**Comment by vito** [comment]
@drahnr Hmm I think something's a bit off with the merge/squash/rebase/whatever. There are a bunch of commits rather than just the one. No big deal though, we can squash + merge when the time comes.

The build failed but it's our fault; fixing it now. I'll re-run it when things are stable.

**Comment by drahnr** [comment]
I did just a regular rebase, but I guess the merge conflicts (for whatever reason there were any)  caused this.

File: assets/common.sh [hunk 3]
```diff
@@ -57,12 +57,14 @@ start_docker() {
...
 
-  for registry in $1; do
+  server_args="${server_args} --max-concurrent-downloads=$1 --max-concurrent-uploads=$2"
+
+  for registry in $3; do
     server_args="${server_args} --insecure-registry ${registry}"
   done
 
-  if [ -n "$2" ]; then
-    server_args="${server_args} --registry-mirror $2"
```
Comment (vito) [inline assets/common.sh]: Sadly the '=' here causes the build failure. Could you replace it with a space?

**Comment by vito** [comment]
Sweet, thanks!

**Pull request merged**